sw/device/silicon_creator/lib/drivers/spi_device.c

```diff
@@ -0,0 +1,66 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/spi_device.h"
+
+#include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "spi_device_regs.h"  // Generated.
+
+enum {
+  /**
+   * Base address of the spi_device registers.
+   */
+  kBase = TOP_EARLGREY_SPI_DEVICE_BASE_ADDR,
+};
+
+void spi_device_init(void) {
+  // CPOL = 0, CPHA = 0, MSb-first TX and RX, 3-byte addressing.
+  uint32_t reg = bitfield_bit32_write(0, SPI_DEVICE_CFG_CPOL_BIT, false);
+  reg = bitfield_bit32_write(reg, SPI_DEVICE_CFG_CPHA_BIT, false);
+  reg = bitfield_bit32_write(reg, SPI_DEVICE_CFG_TX_ORDER_BIT, false);
+  reg = bitfield_bit32_write(reg, SPI_DEVICE_CFG_RX_ORDER_BIT, false);
+  reg = bitfield_field32_write(reg, SPI_DEVICE_CFG_TIMER_V_FIELD, 0x7f);
+  reg = bitfield_bit32_write(reg, SPI_DEVICE_CFG_ADDR_4B_EN_BIT, false);
+  reg = bitfield_bit32_write(reg, SPI_DEVICE_CFG_MAILBOX_EN_BIT, false);
+  abs_mmio_write32(kBase + SPI_DEVICE_CFG_REG_OFFSET, reg);
+
+  // JEDEC manufacturer and device ID.
+  // spi_device sends these in the following order: continuation codes,
+  // manufacturer ID, LSB of device ID, and MSB of device ID.
+  reg = bitfield_field32_write(0, SPI_DEVICE_JEDEC_CC_CC_FIELD,
+                               kSpiDeviceJedecContCode);
+  reg = bitfield_field32_write(reg, SPI_DEVICE_JEDEC_CC_NUM_CC_FIELD,
+                               kSpiDeviceJedecContCodeCount);
+  abs_mmio_write32(kBase + SPI_DEVICE_JEDEC_CC_REG_OFFSET, reg);
+  reg = bitfield_field32_write(
+      0, SPI_DEVICE_JEDEC_ID_ID_FIELD,
+      kSpiDeviceJedecDensity << 8 | kSpiDeviceJedecDevId);
+  reg = bitfield_field32_write(reg, SPI_DEVICE_JEDEC_ID_MF_FIELD,
+                               kSpiDeviceJedecManufId);
+  abs_mmio_write32(kBase + SPI_DEVICE_JEDEC_ID_REG_OFFSET, reg);
+
+  // Configure READ_JEDEC_ID command (CMD_INFO_3).
+  reg = bitfield_field32_write(0, SPI_DEVICE_CMD_INFO_3_OPCODE_3_FIELD,
+                               kSpiDeviceCmdReadJedecId);
+  reg = bitfield_bit32_write(reg, SPI_DEVICE_CMD_INFO_3_VALID_3_BIT, true);
+  abs_mmio_write32(kBase + SPI_DEVICE_CMD_INFO_3_REG_OFFSET, reg);
```

Comment:
Is this really necessary?  Reading the jedec ID should be automatically handled by the spi hardware.

Reply:
Yep. At least valid and opcode are needed in the Flash mode. In passthrough mode, more fields are needed.

Reply:
This surprised me when I first looked at it, too, but there are fixed-function cmd_info slots that represent the hardware-handled features. So, it only _optionally_ handles Read Identification in hardware.